virtio-devices/src/balloon.rs

```diff
@@ -94,6 +95,59 @@ pub struct VirtioBalloonConfig {
     actual: u32,
 }
 
+#[derive(Clone, Debug)]
+struct PartiallyBalloonedPage {
+    addr: u64,
+    bitmap: Vec<u64>,
+}
+
+impl PartiallyBalloonedPage {
+    fn new() -> Self {
+        let size = get_page_size();
+        let len = ((size >> VIRTIO_BALLOON_PFN_SHIFT) + 63) / 64;
+        Self {
+            addr: 0,
+            bitmap: vec![0; len as usize],
+        }
+    }
+
+    fn pfn_match(&self, addr: u64) -> bool {
+        let page_size: u64 = get_page_size();
+        self.addr == addr & !(page_size - 1)
+    }
+
+    fn bitmap_full(&mut self, page_size: u64) -> bool {
+        let mut bp_num_per_page = page_size >> VIRTIO_BALLOON_PFN_SHIFT;
+        for b in &self.bitmap {
+            if bp_num_per_page <= 64 {
+                return b & ((1 << bp_num_per_page) - 1) == (1 << bp_num_per_page) - 1;
+            } else {
+                if *b != u64::MAX {
+                    return false;
+                }
+                bp_num_per_page -= 64;
+            }
+        }
+
+        true
+    }
+
+    fn set_bit(&mut self, page_size: u64, addr: u64) {
+        if !self.pfn_match(addr) {
+            return;
+        }
+        let addr_offset = (addr % page_size) >> VIRTIO_BALLOON_PFN_SHIFT;
+        self.bitmap[(addr_offset / 64) as usize] |= 1 << (addr_offset % 64);
+    }
+
+    fn reset(&mut self) {
+        let size = get_page_size();
+        let len = ((size >> VIRTIO_BALLOON_PFN_SHIFT) + 63) / 64;
+        self.addr = 0;
+        self.bitmap = vec![0; len as usize];
```

Comment:
Ditto: it's better to pre-initialize padding bits as `1`.

Reply:
OK.

Reply:
I believe you forgot to reset the padding bits to `1` as you did from `new()`.